pkg/rate/limiter.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rate
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// Limiter is used to limit the number of operations done.
+type Limiter struct {
+	semaphore   *semaphore.Weighted
+	currWeights int64
+	ticker      *time.Ticker
+	cancelFunc  context.CancelFunc
+	ctx         context.Context
+}
+
+// NewLimiter returns a new Limiter that allows events up to b tokens during
+// the given interval.
+func NewLimiter(interval time.Duration, b int64) *Limiter {
+	ticker := time.NewTicker(interval)
+	ctx, cancel := context.WithCancel(context.Background())
+	l := &Limiter{
+		semaphore:   semaphore.NewWeighted(b),
+		ticker:      ticker,
+		currWeights: 0,
+		ctx:         ctx,
+		cancelFunc:  cancel,
+	}
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+			case <-l.ctx.Done():
+				return
+			}
+			currWeights := atomic.LoadInt64(&l.currWeights)
+			atomic.AddInt64(&l.currWeights, -currWeights)
+			l.semaphore.Release(currWeights)
+		}
+	}()
+	return l
+}
+
+// Stop stops the internal components used for the rate limiter logic.
+func (lim *Limiter) Stop() {
+	lim.cancelFunc()
+	lim.ticker.Stop()
+}
+
+func (lim *Limiter) assertAlive() {
+	select {
+	case <-lim.ctx.Done():
+		panic("limiter misuse: Allow / Wait / WaitN called concurrently after Stop")
+	default:
+	}
+}
+
+// Allow is shorthand for AllowN(time.Now(), 1).
+func (lim *Limiter) Allow() bool {
+	return lim.AllowN(1)
+}
+
+// AllowN returns true if it's possible to allow n tokens.
+func (lim *Limiter) AllowN(n int64) bool {
+	lim.assertAlive()
+	acq := lim.semaphore.TryAcquire(n)
+	if acq {
+		atomic.AddInt64(&lim.currWeights, n)
+		return true
+	}
+	return false
+}
+
+// Wait is shorthand for WaitN(ctx, 1).
+func (lim *Limiter) Wait(ctx context.Context) error {
+	return lim.WaitN(ctx, 1)
+}
+
+// WaitN acquires n tokens, blocking until resources are available or ctx is
+// done. On success, returns nil. On failure, returns ctx.Err() and leaves the
+// limiter unchanged.
+//
+// If ctx is already done, WaitN may still succeed without blocking.
+func (lim *Limiter) WaitN(ctx context.Context, n int64) error {
```

Comment:
So if `n > b`, then this will get stuck forever (or until the context is cancelled), which seems dangerous. This is different from `x/time/rate` which will just return an error instead if `n` is larger than the burst size `b`.

We should either also return an error or just not export `WaitN`.

Reply:
Good point! will do